vpr/src/place/simpleRL_move_generator.cpp

```diff
@@ -0,0 +1,408 @@
+    avail_moves.push_back(std::move(feasible_region_move_generator));
+
+    karmed_bandit_agent = std::move(agent);
+}
+
+//SimpleRL class member functions
+SimpleRLMoveGenerator::SimpleRLMoveGenerator(std::unique_ptr<EpsilonGreedyAgent>& agent) {
+    //SimpleRLMoveGenerator::SimpleRLMoveGenerator(std::unique_ptr<SoftmaxAgent>& agent){
+
+    std::unique_ptr<MoveGenerator> move_generator;
+    move_generator = std::make_unique<UniformMoveGenerator>();
+    avail_moves.push_back(std::move(move_generator));
+
+    std::unique_ptr<MoveGenerator> median_move_generator;
+    median_move_generator = std::make_unique<MedianMoveGenerator>();
+    avail_moves.push_back(std::move(median_move_generator));
+
+    std::unique_ptr<MoveGenerator> weighted_centroid_move_generator;
+    weighted_centroid_move_generator = std::make_unique<WeightedCentroidMoveGenerator>();
+    avail_moves.push_back(std::move(weighted_centroid_move_generator));
+
+    std::unique_ptr<MoveGenerator> centroid_move_generator;
+    centroid_move_generator = std::make_unique<CentroidMoveGenerator>();
+    avail_moves.push_back(std::move(centroid_move_generator));
+
+    std::unique_ptr<MoveGenerator> weighted_median_move_generator;
+    weighted_median_move_generator = std::make_unique<WeightedMedianMoveGenerator>();
+    avail_moves.push_back(std::move(weighted_median_move_generator));
+
+    std::unique_ptr<MoveGenerator> crit_uniform_move_generator;
+    crit_uniform_move_generator = std::make_unique<CriticalUniformMoveGenerator>();
+    avail_moves.push_back(std::move(crit_uniform_move_generator));
+
+    std::unique_ptr<MoveGenerator> feasible_region_move_generator;
+    feasible_region_move_generator = std::make_unique<FeasibleRegionMoveGenerator>();
+    avail_moves.push_back(std::move(feasible_region_move_generator));
+
+    karmed_bandit_agent = std::move(agent);
+}
+
+e_create_move SimpleRLMoveGenerator::propose_move(t_pl_blocks_to_be_moved& blocks_affected, float rlim, std::vector<int>& X_coord, std::vector<int>& Y_coord, e_move_type& move_type, const t_placer_opts& placer_opts, const PlacerCriticalities* criticalities) {
+    move_type = (e_move_type)karmed_bandit_agent->propose_action();
+    return avail_moves[(int)move_type]->propose_move(blocks_affected, rlim, X_coord, Y_coord, move_type, placer_opts, criticalities);
+}
+
+void SimpleRLMoveGenerator::process_outcome(double reward, int reward_num) {
+    karmed_bandit_agent->process_outcome(reward, reward_num);
+}
+/*                                  *
+ *                                  *
+ *  Softmax agent implementation    *
+ *                                  *
+ *                                  */
+
+SoftmaxAgent::SoftmaxAgent(size_t k) {
+    set_k(k);
+    set_action_prob();
+
+    //f_ = vtr::fopen("agent_info.txt", "w");
+}
+
+SoftmaxAgent::~SoftmaxAgent() {
+    if (f_) vtr::fclose(f_);
+}
+
+void SoftmaxAgent::process_outcome(double reward, int reward_num) {
```

Comment:
Duplicated code.

Reply:
removed it by moving this method and the data member used by it to the base class